cpp/include/cudf/detail/gather.cuh

```diff
@@ -518,6 +506,55 @@ void gather_bitmask(table_view const& source,
   }
 }
 
+template <typename MapItRoot>
+struct column_gatherer_impl<struct_view, MapItRoot> {
+  std::unique_ptr<column> operator()(column_view const& column,
+                                     MapItRoot gather_map_begin,
+                                     MapItRoot gather_map_end,
+                                     bool nullify_out_of_bounds,
+                                     cudaStream_t stream,
+                                     rmm::mr::device_memory_resource* mr)
+  {
+    structs_column_view structs_column(column);
+    auto gather_map_size{std::distance(gather_map_begin, gather_map_end)};
+    if (gather_map_size == 0) { return empty_like(column); }
+
+    std::vector<std::unique_ptr<cudf::column>> output_struct_members;
+    std::transform(structs_column.child_begin(),
+                   structs_column.child_end(),
+                   std::back_inserter(output_struct_members),
+                   [&gather_map_begin, &gather_map_end, nullify_out_of_bounds, stream, mr](
+                     cudf::column_view const& col) {
+                     return cudf::type_dispatcher(col.type(),
+                                                  column_gatherer{},
+                                                  col,
+                                                  gather_map_begin,
+                                                  gather_map_end,
+                                                  nullify_out_of_bounds,
+                                                  stream,
+                                                  mr);
+                   });
+
+    gather_bitmask(
+      // Table view of struct column.
+      cudf::table_view{
+        std::vector<cudf::column_view>{structs_column.child_begin(), structs_column.child_end()}},
+      gather_map_begin,
+      output_struct_members,
+      nullify_out_of_bounds ? gather_bitmask_op::NULLIFY : gather_bitmask_op::DONT_CHECK,
+      mr,
+      stream);
+
+    return cudf::make_structs_column(
+      gather_map_size,
+      std::move(output_struct_members),
+      0,
+      rmm::device_buffer{0, stream, mr},  // Null mask will be fixed up in cudf::gather().
```

Comment:
Doesn't the top-level gather logic expect the null mask to be pre-allocated? i.e., it won't allocate one for you?

Reply:
> Doesn't the top-level gather logic expect the null mask to be pre-allocated?

Apparently not:
https://github.com/rapidsai/cudf/blob/d5606afac82aeeae08daba1a76750649b930c68f/cpp/include/cudf/detail/gather.cuh#L475